ufl/cell.py

```diff
@@ -1,298 +1,456 @@
+            if s != o:
+                return s < o
+            return self._lt(other)
+        else:
+            if type(self).__name__ == type(other).__name__:
+                raise ValueError("Cannot order cell types with the same name")
+            return type(self).__name__ < type(other).__name__
 
-class AbstractCell(object):
-    """Representation of an abstract finite element cell with only the
-    dimensions known.
+    def num_vertices(self) -> int:
+        """Get the number of vertices."""
+        return self.num_sub_entities(0)
 
-    """
-    __slots__ = ("_topological_dimension",
-                 "_geometric_dimension")
+    def num_edges(self) -> int:
+        """Get the number of edges."""
+        return self.num_sub_entities(1)
 
-    def __init__(self, topological_dimension, geometric_dimension):
-        # Validate dimensions
-        if not isinstance(geometric_dimension, numbers.Integral):
-            raise ValueError("Expecting integer geometric_dimension.")
-        if not isinstance(topological_dimension, numbers.Integral):
-            raise ValueError("Expecting integer topological_dimension.")
-        if topological_dimension > geometric_dimension:
-            raise ValueError("Topological dimension cannot be larger than geometric dimension.")
+    def num_faces(self) -> int:
+        """Get the number of faces."""
+        return self.num_sub_entities(2)
 
-        # Store validated dimensions
-        self._topological_dimension = topological_dimension
-        self._geometric_dimension = geometric_dimension
-
-    def topological_dimension(self):
-        "Return the dimension of the topology of this cell."
-        return self._topological_dimension
-
-    def geometric_dimension(self):
-        "Return the dimension of the space this cell is embedded in."
-        return self._geometric_dimension
-
-    def is_simplex(self):
-        "Return True if this is a simplex cell."
-        raise NotImplementedError("Implement this to allow important checks and optimizations.")
-
-    def has_simplex_facets(self):
-        "Return True if all the facets of this cell are simplex cells."
-        raise NotImplementedError("Implement this to allow important checks and optimizations.")
-
-    def __lt__(self, other):
-        "Define an arbitrarily chosen but fixed sort order for all cells."
-        if not isinstance(other, AbstractCell):
-            return NotImplemented
-        # Sort by gdim first, tdim next, then whatever's left
-        # depending on the subclass
-        s = (self.geometric_dimension(), self.topological_dimension())
-        o = (other.geometric_dimension(), other.topological_dimension())
-        if s != o:
-            return s < o
-        return self._ufl_hash_data_() < other._ufl_hash_data_()
+    def num_volumes(self) -> int:
+        """Get the number of faces."""
+        return self.num_sub_entities(3)
```

Comment:
Subtly wrong if 4-D and higher cells are ever implemented. Should probably be `self.num_sub_entities(tdim)`.